### PR TITLE
Mohammad moving adaptive X-over-n-mutation

### PR DESCRIPTION
**Pull request by Jimmy-INL**
--------
Pull Request Description
--------
##### What are the significant changes in functionality due to this change request?

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [ ] 1. Review all computer code.
- [ ] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [ ] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [ ] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [ ] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [ ] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [ ] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [ ] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [ ] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: dependencies.xml [hunk 1]
```diff
@@ -84,6 +84,8 @@ Note all install methods after "main" take
     <cyipopt optional='True'/>
     <pyomo-extensions source="pyomo" skip_check='True' optional='True'/>
     <setuptools />
+    <!-- This is because liblapack 3.9.0 build 21 is broken (and can probably be removed if there ever is a build 22).  This can also be removed when scipy is updated to version 1.12 -->
+    <liblapack skip_check='True' os='windows' machine='x86_64'>3.9.0=20_win64_mkl</liblapack>
```
Comment (wangcj05) [inline dependencies.xml]: We do not have these lines in our devel now.

File: ravenframework/Optimizers/BayesianOptimizer.py [hunk 2]
```diff
@@ -150,6 +150,7 @@ def __init__(self):
     self._paramSelectionOptions = {'ftol':1e-10, 'maxiter':200, 'disp':False} # Optimizer options for hyperparameter selection
     self._externalParamOptimizer = 'fmin_l_bfgs_b'                            # Optimizer for external hyperparameter selection
     self._resetModel = False                                                  # Reset regression model if True
+    self._canHandleMultiObjective = False                         # boolean indicator whether optimization is a sinlge-objective problem or a multi-objective problem
```
Comment (wangcj05) [inline ravenframework/Optimizers/BayesianOptimizer.py]: This is in Optimizer.py, and you do not need to specify it here.

File: ravenframework/Optimizers/GradientDescent.py [hunk 3]
```diff
@@ -212,6 +212,7 @@ def __init__(self):
     self._followerProximity = 1e-2   # distance at which annihilation can start occurring, in ?normalized? space
     self._trajectoryFollowers = defaultdict(list) # map of trajectories to the trajectories following them
     self._functionalConstraintExplorationLimit = 500 # number of input-space explorations allowable for functional constraints
+    self._canHandleMultiObjective = False # Currently Gradient Descent cannot handle multiobjective optimization
```
Comment (wangcj05) [inline ravenframework/Optimizers/GradientDescent.py]: I think you do not need to specify it here since it is presented in the base class.

File: ravenframework/Optimizers/SimulatedAnnealing.py [hunk 4]
```diff
@@ -191,6 +191,7 @@ def __init__(self):
     self._coolingMethod = None                                  # initializing cooling method
     self._coolingParameters = {}                                # initializing the cooling schedule parameters
     self.info = {}
+    self._canHandleMultiObjective = False                       # Currently Simulated Annealing can only handle single objective
```
Comment (wangcj05) [inline ravenframework/Optimizers/SimulatedAnnealing.py]: you do not need to add it here.

File: ravenframework/Optimizers/crossOverOperators/crossovers.py [hunk 5]
```diff
@@ -153,6 +153,26 @@ def twoPointsCrossover(parents, **kwargs):
 
   return children
 
+def getLinearCrossoverProbability(iter, limit):
+  """
+  This method is designed to ILC(Increasing Low Crossover) adaptive crossover methodology each iteration with probability.
+  @ In, Current iteration number, Total iteration number
+  @ Out, (iteration / limit) as crossover rate
```
Comment (wangcj05) [inline ravenframework/Optimizers/crossOverOperators/crossovers.py]: please update the docstring here to align with raven docstring standards

File: ravenframework/Optimizers/crossOverOperators/crossovers.py [hunk 6]
```diff
@@ -153,6 +153,26 @@ def twoPointsCrossover(parents, **kwargs):
...
+  """
+  return iter/limit
+
+def getQuadraticCrossoverProbability(iter, limit):
+  """
+  This method is designed to quadratic adaptive crossover methodology each iteration with probability.
+  @ In, Current iteration number, Total iteration number
+  @ Out,  (iteration+1/limit)^2 as crossover rate
```
Comment (wangcj05) [inline ravenframework/Optimizers/crossOverOperators/crossovers.py]: please update the docstring here to align with raven docstring standards

File: ravenframework/Optimizers/crossOverOperators/crossovers.py [hunk 7]
```diff
@@ -153,6 +153,26 @@ def twoPointsCrossover(parents, **kwargs):
...
+  """
+  if(iter == 0):
+    crossoverProb = 0
+  else:
+    crossoverProb = ((iter+1)/(limit))**2
+  return crossoverProb
```
Comment (wangcj05) [inline ravenframework/Optimizers/crossOverOperators/crossovers.py]: Any reason why there is inconsistency when calculating crossoverProb?

File: ravenframework/Optimizers/fitness/fitness.py [hunk 8]
```diff
@@ -187,6 +186,8 @@ def logistic(rlz, **kwargs):
 __fitness['invLinear'] = invLinear
 __fitness['logistic']  = logistic
 __fitness['feasibleFirst'] = feasibleFirst
+#NOTE hardConstraint method will be used later once constraintHandling is realized. Until then, it will be commented. @JunyungKim
+# __fitness['hardConstraint'] = hardConstraint
```
Comment (wangcj05) [inline ravenframework/Optimizers/fitness/fitness.py]: remove?

File: ravenframework/Optimizers/mutators/mutators.py [hunk 9]
```diff
@@ -64,7 +64,6 @@ def scrambleMutator(offSprings, distDict, **kwargs):
     This method performs the scramble mutator. For each child, a subset of genes is chosen
     and their values are shuffled randomly.
     @ In, offSprings, xr.DataArray, offsprings after crossover
-    @ In, distDict, dict, dictionary containing distribution associated with each gene
```
Comment (wangcj05) [inline ravenframework/Optimizers/mutators/mutators.py:67]: why all docstrings about district are removed?

File: ravenframework/Optimizers/mutators/mutators.py [hunk 10]
```diff
@@ -170,7 +166,7 @@ def inversionMutator(offSprings, distDict, **kwargs):
       seq = np.arange(locL,locU+1)
       allElems = []
       for i,elem in enumerate(seq):
-        allElems.append(distDict[child.coords['Gene'].values[i]].cdf(float(child[elem].values)))
+         allElems.append(distDict[child.coords['Gene'].values[i]].cdf(float(child[elem].values)))
```
Comment (wangcj05) [inline ravenframework/Optimizers/mutators/mutators.py]: The change need to be reverted here.

File: ravenframework/Optimizers/mutators/mutators.py [hunk 11]
```diff
@@ -195,6 +191,26 @@ def locationsGenerator(offSprings,locs):
   loc2 = np.maximum(locs[0], locs[1])
   return loc1, loc2
 
+def getLinearMutationProbability(iter, limit):
+  """
+  This method is designed to DHM(Decreasing High Mutation) adaptive mutation methodology each iteration with probability.
+  @ In, Current iteration number, Total iteration number
+  @ Out, 1-(iteration / limit) as mutation rate
```
Comment (wangcj05) [inline ravenframework/Optimizers/mutators/mutators.py]: please update the docstring

File: ravenframework/Optimizers/mutators/mutators.py [hunk 12]
```diff
@@ -195,6 +191,26 @@ def locationsGenerator(offSprings,locs):
...
+  """
+  return 1-(iter/limit)
+
+def getQuadraticMutationProbability(iter, limit):
+  """
+  This method is designed to Quadratic adaptive mutation methodology each iteration with probability.
+  @ In, Current iteration number, Total iteration number
+  @ Out, 1-(((1+iteration)/limit))^2 as mutation rate
```
Comment (wangcj05) [inline ravenframework/Optimizers/mutators/mutators.py]: please update the docstring

File: tests/framework/Optimizers/GeneticAlgorithms/discrete/constrained/MinwoRepMultiObjectiveFeasibleFirst.xml [hunk 13]
```diff
@@ -11,6 +12,13 @@
     The design/serach variables are x1, through x3. Each of these variables are integers between 2, and 4.
     The problem is subjected to constraints: expConstr1, ixpConstr2, ixpConstr3 found at myCOnstraints.py
     </description>
+=======
+    <name>raven/tests/framework/Optimizers/GeneticAlgorithms.NSGAII</name>
+    <author>Mohammad Abdo</author>
+    <created>2024-04-27</created>
+    <classesTested/>
+    <description>NSGA-II min-min test </description>
+>>>>>>> 3c6da212951e24c9ceb952e8fbe85829436d8e17
```
Comment (wangcj05) [inline tests/framework/Optimizers/GeneticAlgorithms/discrete/constrained/MinwoRepMultiObjectiveFeasibleFirst.xml]: The merge conflicts are not resolved in this file, please resolve it.

Comment (Jimmy-INL) [reply tests/framework/Optimizers/GeneticAlgorithms/discrete/constrained/MinwoRepMultiObjectiveFeasibleFirst.xml]: Why doesn't it show conflicts down? I am sure I have addressed that, let me check again

File: ravenframework/Optimizers/mutators/mutators.py (hunk 9 above)
Comment (Jimmy-INL) [reply ravenframework/Optimizers/mutators/mutators.py:67]: Done

File: ravenframework/Optimizers/mutators/mutators.py (hunk 12 above)
Comment (Jimmy-INL) [reply ravenframework/Optimizers/mutators/mutators.py]: Done

File: tests/framework/Optimizers/GeneticAlgorithms/discrete/constrained/MinwoRepMultiObjectiveFeasibleFirst.xml (hunk 13 above)
Comment (Jimmy-INL) [reply tests/framework/Optimizers/GeneticAlgorithms/discrete/constrained/MinwoRepMultiObjectiveFeasibleFirst.xml]: Done

File: ravenframework/Optimizers/fitness/fitness.py (hunk 8 above)
Comment (Jimmy-INL) [reply ravenframework/Optimizers/fitness/fitness.py]: done.

File: ravenframework/Optimizers/crossOverOperators/crossovers.py (hunk 6 above)
Comment (Jimmy-INL) [reply ravenframework/Optimizers/crossOverOperators/crossovers.py]: I have updated the dockstrings and removed the `+1 ` this can result in a proabability not in 0< prob<1

File: ravenframework/Optimizers/GeneticAlgorithm.py [hunk 14]
```diff
@@ -663,9 +667,13 @@ def handleInput(self, paramInput):
       self._crossoverPoints = None
     else:
       self._crossoverPoints = crossoverNode.findFirst('points').value
+    crossoverProbNode = crossoverNode.findFirst('crossoverProb')
+    try:
+      self._crossoverProbType = crossoverProbNode.parameterValues['type']
+    except:
+      self._crossoverProbType = 'static'
```
Comment (wangcj05) [inline ravenframework/Optimizers/GeneticAlgorithm.py:674]: First, there is no check for accepted 'type' here. Second, I may suggest to use InputData to handle it. Third, the getInputSpecs need to be updated to reflect crossoverProbType, right now only 'static' and 'adaptive' is presented. In your enhancement, you provide the options static, linear and quadratic.

File: ravenframework/Optimizers/GeneticAlgorithm.py [hunk 15]
```diff
@@ -675,6 +683,11 @@ def handleInput(self, paramInput):
       self._mutationLocs = None
     else:
       self._mutationLocs = mutationNode.findFirst('locs').value
+    mutationProbNode = mutationNode.findFirst('mutationProb')
+    try:
+      self._mutationProbType = mutationProbNode.parameterValues['type']
+    except:
+      self._mutationProbType = 'static'
```
Comment (wangcj05) [inline ravenframework/Optimizers/GeneticAlgorithm.py:690]: Similar comments to crossover part. 
First, there is no check for accepted 'type' here. Second, I may suggest to use InputData to handle it. Third, the getInputSpecs need to be updated to reflect mutationProbType, right now only 'static' and 'adaptive' is presented. In your enhancement, you provide the options static, linear and quadratic.

File: ravenframework/Optimizers/crossOverOperators/crossovers.py [hunk 16]
```diff
@@ -153,6 +153,29 @@ def twoPointsCrossover(parents, **kwargs):
...
+  """
+  This method is designed to quadratic adaptive crossover methodology each iteration with probability.
+  @ In, iter, integer, current iteration number
+  @ In, limit, integer, total number of iterations required
+  @ Out, crossoverProb, float, the linearly adaptive crossover probability
+  """
+  if(iter == 0):
+    crossoverProb = 0
+  else:
+    crossoverProb = ((iter)/(limit))**2
```
Comment (wangcj05) [inline ravenframework/Optimizers/crossOverOperators/crossovers.py:176]: Sorry for the confusion here. My original comment is about why there is inconsistency about the crossoverProb calculation. The new changes make it consistent. If this is what you need, I guess you can remove the if .. else statement here.

File: ravenframework/Optimizers/mutators/mutators.py [hunk 17]
```diff
@@ -195,6 +191,29 @@ def locationsGenerator(offSprings,locs):
...
+  """
+  This method is designed to Quadratic adaptive mutation methodology each iteration with probability.
+  @ In, iter, integer, current iteration number,
+  @ In, limit, integer, total iteration number
+  @ Out, mutationProb, float, the quadratic adaptive mutation probability
+  """
+  if(iter == 0):
+    mutationProb = 1
+  else:
+    mutationProb = 1-(((iter)/(limit))**2)
```
Comment (wangcj05) [inline ravenframework/Optimizers/mutators/mutators.py:214]: same here: Sorry for the confusion here. My original comment is about why there is inconsistency about the crossoverProb calculation. The new changes make it consistent. If this is what you need, I guess you can remove the if .. else statement here.

File: tests/framework/Optimizers/GeneticAlgorithms/continuous/unconstrained/ZDT1-adaptive-LinearXover-LinearMutation.xml [hunk 18]
```diff
@@ -0,0 +1,195 @@
+<?xml version="1.0" ?>
+<Simulation verbosity="debug" profile="jobs">
+  <TestInfo>
+    <name>raven/tests/framework/Optimizers/GA.MultiObjZDT1</name>
+    <author>Mohammad Abdo</author>
+    <created>2023-02-21</created>
```
Comment (wangcj05) [inline tests/framework/Optimizers/GeneticAlgorithms/continuous/unconstrained/ZDT1-adaptive-LinearXover-LinearMutation.xml:6]: Do you want to update the date here?

File: tests/framework/Optimizers/GeneticAlgorithms/continuous/unconstrained/ZDT1-adaptive-LinearXover-LinearMutation.xml [hunk 19]
```diff
@@ -0,0 +1,195 @@
+<?xml version="1.0" ?>
+<Simulation verbosity="debug" profile="jobs">
+  <TestInfo>
+    <name>raven/tests/framework/Optimizers/GA.MultiObjZDT1</name>
```
Comment (wangcj05) [inline tests/framework/Optimizers/GeneticAlgorithms/continuous/unconstrained/ZDT1-adaptive-LinearXover-LinearMutation.xml:4]: You need to update the name "MultiObjZDT" to reflect the real test name in tests. Please do it for the tests, right now they are using the same name.

File: tests/framework/Optimizers/GeneticAlgorithms/continuous/unconstrained/ZDT1.xml [hunk 20]
```diff
@@ -24,7 +24,7 @@
       $0 \leq obj_1 \leq 1$
       $obj_2$
       $1 \geq obj_1 \geq 0$
-
+    \begin{verbatim}
```
Comment (wangcj05) [inline tests/framework/Optimizers/GeneticAlgorithms/continuous/unconstrained/ZDT1.xml:27]: I may suggest removing this file.

File: tests/framework/Optimizers/GeneticAlgorithms/tests [hunk 21]
```diff
@@ -375,59 +375,130 @@
...
-      type = UnorderedCSV
-      output = 'continuous/unconstrained/ZDT1/opt_export_0.csv'
-      rel_err = 0.001
+     type = UnorderedCSV
+     output = 'continuous/unconstrained/ZDT1-static-0.0Xover-1.0Mutation/opt_export_0.csv'
+     rel_err = 0.001
+    [../]
+  [../]
+
+  [./ZDT1-static-0.1Xover-0.0Mutation]
```
Comment (wangcj05) [inline tests/framework/Optimizers/GeneticAlgorithms/tests:440]: I mean this name need to be used in TestInfo of each test.

File: tests/framework/Optimizers/GeneticAlgorithms/continuous/unconstrained/ZDT1-static-1.0Xover-0.0Mutation.xml [hunk 22]
```diff
@@ -0,0 +1,195 @@
+<?xml version="1.0" ?>
+<Simulation verbosity="debug" profile="jobs">
```
Comment (wangcj05) [inline tests/framework/Optimizers/GeneticAlgorithms/continuous/unconstrained/ZDT1-static-1.0Xover-0.0Mutation.xml:2]: Another suggestion for all the tests, you may want to leverage external xml to avoid duplicates of tests with minor changes.

File: ravenframework/Optimizers/crossOverOperators/crossovers.py (hunk 16 above)
Comment (Jimmy-INL) [reply ravenframework/Optimizers/crossOverOperators/crossovers.py:176]: Congjian this was initiated by an intern, his equations were not correct as it could result in probability bigger than 1. I think now it should be better.

File: tests/framework/Optimizers/GeneticAlgorithms/continuous/unconstrained/ZDT1.xml (hunk 20 above)
Comment (Jimmy-INL) [reply tests/framework/Optimizers/GeneticAlgorithms/continuous/unconstrained/ZDT1.xml:27]: without this there was a failure. @joshua-cogliati-inl suggested this if I recall correctly. Or maybe he suggested removing it! @joshua-cogliati-inl please confirm once you can take a look.

Comment (joshua-cogliati-inl) [reply tests/framework/Optimizers/GeneticAlgorithms/continuous/unconstrained/ZDT1.xml:27]: @wangcj05 Which file are you suggesting removing?  Without adding verbatim, we get some errors when building documentation.
